### PR TITLE
Resolve symbol types in symbol table

### PR DESCRIPTION
**Pull request by danpoe**
This resolves the type symbols to the pointed-to types in the symbol table before creating the goto program, when reading a json goto program file.

### CODE REVIEW

File: src/json-symtab-language/json_symtab_language.cpp [hunk 1]
```diff
@@ -39,6 +41,65 @@ bool json_symtab_languaget::typecheck(
   }
 }
 
+void json_symtab_languaget::follow_symbol_types(
+  irept &irep,
+  const namespacet &ns)
+{
+  if(irep.id() == ID_symbol)
```
Comment (chrisr-diffblue) [inline src/json-symtab-language/json_symtab_language.cpp]: Can't this whole if block be replaced with a single call to ns.follow_symbol(irep) ?

**Review by chrisr-diffblue — COMMENTED** [review]
Also, we should figure out how to synchronize this branch with @smowton's original branch....

File: src/json-symtab-language/json_symtab_language.cpp (hunk 1 above)
Comment (danpoe) [reply src/json-symtab-language/json_symtab_language.cpp]: The `follow_symbol()` method follows both type symbols and non-type symbols, but we only want to replace type symbols. So we first do the check `if(symbol.is_type)` and only do the follow in this case.

Comment (chrisr-diffblue) [reply src/json-symtab-language/json_symtab_language.cpp]: Ah, right - got it... A couple of observations:

1) This should probably be renamed to `follow_type_symbols` to match the naming of `follow_symbols`
2) Should either this `follow_type_symbols` function be moved into `namespace.cpp` as a utility function there, or should ns.follow_symbol() grow an extra argument (with a default value) that specifies whether to follow non-type symbols or not, and then just use that where we need it instead?

Comment (danpoe) [reply src/json-symtab-language/json_symtab_language.cpp]: Ah yes. I think I'll move the new `follow_type_symbols` function to `namespace.cpp`.

**Pull request merged**